Move pipeline implementation to a separate object

happz opened the pull request:
A pipeline can spawn a new, derived pipeline. Glue instance then takes
care of managing them, discovering all registered shared functions and
similar top-level stuff.

Primary motivation was to let modules read pipeline description while in
destroy phase, and let modules spawn pipelines. It proved to be worth
the effort to refactor related code, clearing the interface.

happz committed:
Move pipeline implementation to a separate object

A pipeline can spawn a new, derived pipeline. Glue instance then takes
care of managing them, discovering all registered shared functions and
similar top-level stuff.

Primary motivation was to let modules read pipeline description while in
destroy phase, and let modules spawn pipelines. It proved to be worth
the effort to refactor related code, clearing the interface.

happz commented:
Not fully tested yet, and tests need fixing and refactoring to accommodate new structure, but it is definitely open for a review.

thrix commented:
awesome, I will try ot understand how to initialize a new pipeline, so I can try it out ...

happz commented:
> awesome, I will try ot understand how to initialize a new pipeline, so I can try it out ...

It's simple, things didn't change that much:

```python
glue.run_pipeline(glue.Pipeline([glue.PipelineStep('rules-engine')]))
```

is the same as:

```python
glue.run_modules([glue.PipelineStep('rules-engine')])
```

or, since it's just a single module, one can use:

```python
glue.run_module('rules-engine')
```

Documentation and tests will arrive soon, hopefuly.

happz committed:
Tests and extended documentation

happz committed:
Fix type check issue

happz committed:
Callback step and module

```
def foo(glue, s):
  glue.info(s)

glue.run_modules([
  PipelineStepModule('ansible'),
  PipelineStepCallback('name used for logging', foo, 'this will be logged via glue')
])
```

happz committed:
Fixing tests

mkluson commented on gluetool/glue.py:
```diff
@@ -52,6 +54,22 @@
+# Each `Glue` instance manages one or more pipelines. The main one, requested by the user, can spawn additional
+# "side-car" pipelines, and these can spawn their own children, and so on. Only a single pipeline is being processed
+# at the moment - when a new pipeline is spawned, its parent gives up its time for its kid, and is awoken when the
+# kid finishes. `Glue` keeps a stack of pipelines as thei appear and disappear. There is always a "current pipeline",
```
typo: thei

thrix commented:
Cool, will test this properly PipelineCallbackStep looks awesome

thrix commented:
Works for me!

thrix reviewed (dismissed):
I tested, Used the callback. created separate pipeline. Everything seems fine ...

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+    * :py:meth:`_safe_call` - calls a given callback, returns its return value. Any exception raised by the callback
+      is wrapped by :py:class:`Failure` instance and returned instead of what callback would return.
+    * :py:meth:`_for_each_module` - loop over given list of modules, calling given callback for each of the modules.
+      :py:meth:`_safe_call` is used for the call, makign sure we always have a return value and no exceptions.
```
typo: makign

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+        # type: (Iterable[Module], Callable[..., Optional[Failure]], *Any, **Any) -> Optional[Failure]
+        """
+        For each module in a list, call a given function with module as its first argument. If the call
+        returns anythign but ``None``, the value is returned by this function as well, ending the loop.
```
typo: anythign

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+
+            # Given that we're using `_safe_call`, we shouldn't encounter any exception - `_safe_call`
+            # would convert any exception into `Failure` instance. Callback also cannot return either
+            # `None` or a failure. Therefore, if we gat anything `True`-ish, we simply pass it to our
```
typo: gat

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+            # caller since it must by a failure.
+            #
+            # Note: This is enforced by type checks, we have no other power over the callback and its
+            # return typ.e
```
typo: typ.e

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+        # Make a copy of steps - while setting modules up, we won't have access to module index, so we cannot
+        # reach to `self.steps` for its arguments, but we can pop the first item of this list - it's always
+        # the "current" module, the one currently being set up.
+        steps = self.steps[:]
```
what is the difference between this and implementation and `self.steps.copy()`?

ukulekek replied:
They are equal. Nevertheless, `list.copy()` is not available in python 2.7

happz replied:
Right. Both work the same way, I'm used to `[:]`.

mkluson commented on gluetool/glue.py:
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
+            #
+            # Therefore, we catch the (possible) exception raised by `execute`, save its info aside, register shared
+            # functions, and re-raise the exception. By the act of saving the exception info, we let our exception
+            # chaining to catch up with us - should `add_shared` raise any exception, it would correctly recognize
```
typo: corrently

happz replied:
This I cannot find, looks got to me.

mkluson commented on gluetool/glue.py:
```diff
@@ -1020,68 +1508,87 @@ def parse_args(self, args):
+        # type: (str, *Any, **Any) -> Any
         """
+        Call a shared function overloaded by the one provided by this module. This way,
+        a module can give chance to other implementations of its action, e.g. to publish
```
do not mention message bus here, it is to tied to our usecases

happz replied:
I think it's a legitimate example, we're not the only project in the world sending messages to a message bus.

happz replied:
@mkluson I tried to rewrite the wording slightly, let me know your opinion (or resolve the conversation).

happz replied:
@mkluson so, how are we standing on the review of my update?

happz committed:
Typos and doc tweaks

happz committed:
Fixing tests

mkluson commented on gluetool/tests/test_run_modules.py:
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
+
+def test_for_each_module(pipeline):
+    modules = [
+        create_module(DummyModule, name=DummyModule.name)[1],
```
Can `module` fixture be used here?

happz replied:
No, it must be two different instances (`assert module is expected_module`). I could use `module` for one instance and create the other, it seemed to be simpler to create both of them.

mkluson commented on gluetool/tests/test_run_modules.py:
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
+    assert ret.exception.message == 'bar'
+
+
+def test_pipeline_setup(pipeline, monkeypatch):
```
param `monkeypatch` is unused in this function

happz replied:
Fixed.

mkluson commented on gluetool/tests/test_run_modules.py:
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
+        pipeline.add_shared('does_not_exist', module)
+
+
+def test_has_shared(glue, pipeline, module):
```
param `module` is unused

happz replied:
Fixed.

thrix reviewed (approved):
LGTM, tested, using in almost production :D

thrix commented:
@happz  I believe we can merge this without pull approve .. this requerest have been reviewed as needed ...
